File: bench/programs/mandelbrot/mandelbrot.go
package main

import (
	"fmt"
	"time"
)

func run() int64 {
	var w, h, maxiter int64 = 800, 800, 255
	var checksum int64 = 0

	// Go may contract `a*b + c` into a single FMA, which rounds differently
	// from the other languages' separately-rounded mul-then-add (and the SSA
	// pass re-fuses even across statements). A `float64(...)` conversion on the
	// product is the only reliable barrier: it forces the multiply to round to
	// double before the add, matching everyone else's checksum.
	for py := int64(0); py < h; py++ {
		for px := int64(0); px < w; px++ {
			cre := -2.0 + float64(float64(px)*(2.5/float64(w)))
			cim := -1.25 + float64(float64(py)*(2.5/float64(h)))
			var zr, zi, zr2, zi2 float64 = 0, 0, 0, 0
			var i int64 = 0
			for zr2+zi2 <= 4.0 {
				if i >= maxiter {
					break
				}
				zi = float64(2.0*zr*zi) + cim
				zr = zr2 - zi2 + cre
				zr2 = zr * zr
				zi2 = zi * zi
				i++
			}
			checksum += i
		}
	}
	return checksum
}

func main() {
	run() // warmup
	t0 := time.Now()
	ans := run()
	dt := time.Since(t0).Nanoseconds()
	fmt.Println(ans)
	fmt.Println(dt)
}

File: bench/programs/mandelbrot/mandelbrot.py
import time


def run():
    w = 800
    h = 800
    maxiter = 255
    checksum = 0
    for py in range(h):
        for px in range(w):
            cre = -2.0 + px * (2.5 / w)
            cim = -1.25 + py * (2.5 / h)
            zr = 0.0
            zi = 0.0
            zr2 = 0.0
            zi2 = 0.0
            i = 0
            while zr2 + zi2 <= 4.0:
                if i >= maxiter:
                    break
                zi = 2.0 * zr * zi + cim
                zr = zr2 - zi2 + cre
                zr2 = zr * zr
                zi2 = zi * zi
                i += 1
            checksum += i
    return checksum


run()  # warmup
t0 = time.perf_counter_ns()
ans = run()
dt = time.perf_counter_ns() - t0
print(ans)
print(dt)
